plumbing/ps/testing.go

```diff
@@ -0,0 +1,83 @@
+package ps
+
+import (
+	"context"
+	"gx/ipfs/QmVRxA4J3UPQpw74dLrQ6NJkfysCA1H4GU28gVpXQt9zMU/go-libp2p-pubsub"
+	"sync"
+)
+
+// FakeSubscription is a fake pubsub subscription.
+type FakeSubscription struct {
+	topic       string
+	pending     chan *pubsub.Message
+	err         error
+	cancelled   bool
+	awaitCancel sync.WaitGroup
+}
+
+// NewFakeSubsciption builds a new fake subscription to a topic.
+func NewFakeSubscription(topic string) *FakeSubscription {
+	sub := &FakeSubscription{
+		topic:       topic,
+		pending:     make(chan *pubsub.Message),
+		awaitCancel: sync.WaitGroup{},
+	}
+	sub.awaitCancel.Add(1)
+	return sub
+}
+
+// Subscription interface
+
+// Topic returns this subscription's topic.
+func (s *FakeSubscription) Topic() string {
+	return s.topic
+}
+
+// Next returns the next messages from this subscription.
+func (s *FakeSubscription) Next(ctx context.Context) (*pubsub.Message, error) {
+	if s.err != nil {
+		return nil, s.err
+	}
+	select {
+	case msg := <-s.pending:
```

Comment:
pointing this out now not bc i think it is important here bc it doesnt seem so, but because this is one of the sharp edges of go channels: receiving on a channel actually returns two values. the second is a bool signifying the channel is still open. if the second is false then the channel was closed by caller and the first received value is the zero value for the type. i have been bitten by this many times so make a point of mentioning it to anyone who might not have had the pleasure of hitting it.